### PR TITLE
Routing rework (sc-page-selector)

### PR DESCRIPTION
**Pull request by robertmusil**

### CODE REVIEW

**Comment by sujato** [comment]
I'm running this branch locally, and it is mostly working, but there are a number of bugs. I'm not sure how useful it is to try to iron them all out. Maybe better to merge it first and then fix the bugs? Like I said earlier, the previous behavior was not always correct, so we'll have to work through each case carefully.

One rather curious bug, on the /define dictionary pages, the title appears in ascii not Unicode:

Screenshot from 2021-02-05 21-48-33

**Comment by sujato** [comment]
another bug is that the tab bar doesn't appear on the Home page

File: client/elements/sc-page-selector.js [hunk 1]
```diff
@@ -1,289 +1,508 @@
...
+    }
+  }
+
+  _scrollToTop() {
+    window.scroll(0, 0);
   }
 
   _redirectFromLegacyLink() {
-    let parts = this.route.path.split('/');
+    let parts = this.router.location.pathname.split('/');
```
Comment (Prograd) [inline client/elements/sc-page-selector.js:358]: nit: I think it can be constant.

File: client/elements/sc-page-selector.js [hunk 2]
```diff
@@ -1,289 +1,508 @@
...
+        } else {
+          this.actions.changeToolbarTitle('');
+        }
+    }
+  }
 
-    this._resolveImports();
+  _updateNav() {
+    let navArray = store.getState().navigationArray;
+    let currentPath = this.currentRoute.path;
```
Comment (Prograd) [inline client/elements/sc-page-selector.js:465]: nit: Both variable can be constant. You are not reassign them in the next line.

File: client/elements/sc-page-selector.js [hunk 3]
```diff
@@ -300,208 +519,14 @@ class SCPageSelector extends ReduxMixin(Localized(PolymerElement)) {
...
-    this.dispatchEvent(new CustomEvent('show-sc-toast', {
-      detail: {
-        toastType: toastType,
-        message: text
-      },
-      bubbles: true,
-      composed: true
-    }));
+    this.parentNode.querySelector('#action_items').style.display =
+      this.currentRoute.name === 'HOME' ? 'none' : 'flex';
```
Comment (Prograd) [inline client/elements/sc-page-selector.js:523]: nit: Seems like not format correctly

File: client/elements/sc-site-layout.js [hunk 4]
```diff
@@ -308,50 +432,67 @@ class SCSiteLayout extends LitLocalized(LitElement) {
...
-        rootDOM.getElementById('breadCrumb').style.transform = 'translateY(120px)';
-        rootDOM.getElementById('mainTitle').style.transform = 'translateY(74px) scale(0.667)';
-        rootDOM.getElementById('subTitle').style.opacity = '0';
-        rootDOM.getElementById('subTitle').style.transform = 'scale(0)';
-        if (window.innerWidth < 480) {
-          rootDOM.getElementById('mainTitle').style.transform = 'translateY(70px) scale(0.667)';
+    addEventListener(
+      'scroll',
+      throttle(500, () => {
+        let transitionStyle = 'transform 200ms ease-in-out';
```
Comment (Prograd) [inline client/elements/sc-site-layout.js:438]: nit: it can be constant

File: client/elements/sc-site-layout.js [hunk 5]
```diff
@@ -308,50 +432,67 @@ class SCSiteLayout extends LitLocalized(LitElement) {
...
+        let displaySuttaToC = store.getState().displaySuttaToC;
+        const displaySuttaInfo = store.getState().displaySuttaInfo;
+        if (
+          this.changedRoute.path !== '/' &&
+          !displaySettingMenu &&
+          !displaySuttaParallels &&
+          !displaySuttaInfo &&
+          !displaySuttaToC
+        ) {
+          let currentScrollTop = window.pageYOffset || document.documentElement.scrollTop;
```
Comment (Prograd) [inline client/elements/sc-site-layout.js:485]: nit: all variables can be constant here.

File: client/elements/sc-site-layout.js [hunk 6]
```diff
@@ -402,22 +543,22 @@ class SCSiteLayout extends LitLocalized(LitElement) {
...
-    if (changedProps.has('siteLanguage')) {
-      this._setSiteLanguage();
-    }
-    if (changedProps.has('appColorTheme')) {
-      this._colorThemeChanged();
-    }
-    if (changedProps.has('changedRoute')) {
-      this._routeChanged();
+  updated(changedProps) {
+    //super.updated(changedProps);
```
Comment (Prograd) [inline client/elements/sc-site-layout.js:551]: Remove the comment, it's unnecessary and meaningless.

**Review by Prograd — APPROVED** [review]

**Comment by sujato** [comment]
another bug, the more-menu is showing a console error:

```
sc-universal-action-items.js?60b2:117 Uncaught (in promise) TypeError: Cannot read property 'length' of undefined
    at HTMLElement.firstUpdated (sc-universal-action-items.js?60b2:117)
    at HTMLElement.performUpdate (updating-element.js?b22b:571)
    at HTMLElement._enqueueUpdate (updating-element.js?b22b:507)
```

File: client/utils/routingService.js [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+    ) {
+      return;
+    }
+
+    let url;
+    if (document.baseURI != null) {
+      url = new URL(anchor.href, document.baseURI);
+    } else {
+      url = new URL(anchor.href);
+    }
```
[the remark this answers is not in the document]
Comment (robertmusil) [reply client/utils/routingService.js:118]: because it's too long then : )

File: client/elements/sc-site-layout.js (hunk 4 above)
Comment (robertmusil) [reply client/elements/sc-site-layout.js:438]: lets leave it for now, I only reformatted those components for now, lets not touch this.

File: client/elements/sc-page-selector.js (hunk 3 above)
Comment (robertmusil) [reply client/elements/sc-page-selector.js:523]: Nah it's all prettiered

File: client/elements/sc-site-layout.js (hunk 4 above)
Comment (ihongda) [reply client/elements/sc-site-layout.js:438]: I will gradually modify these codes according to eslint's prompts :)

**Comment by ihongda** [comment]
Clicking on some menus will cause this error: 
深度截图_选择区域_20210206114036

**Comment by ihongda** [comment]
In addition, there are some problems with the display of the breadcrumb items:)

File: client/utils/routingService.js [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+  }
+
+  /**
+   * @param e {MouseEvent}
+   * @private
+   */
+  _anchorHandler(e) {
+    /** @type {HTMLAnchorElement} */
+    let anchor = null;
+    for (let elem of e.path) {
```
[the remark this answers is not in the document]
Comment (robertmusil) [reply client/utils/routingService.js:79]: Actually needs elem reference here.

**Comment by sujato** [comment]
> another bug is that the tab bar doesn't appear on the Home page

This issue is not fixed.

Screenshot from 2021-02-08 20-42-22

**Pull request merged**